ipam/pool_test.go

```diff
@@ -0,0 +1,991 @@
+						Mask: net.IPv6zero.DefaultMask(),
+					}
+					poolId := subnet.String()
+					as := &addressSpace{
+						Id:    "asId",
+						Pools: map[string]*addressPool{},
+					}
+					pool, err := as.newAddressPool("eth0", 1, subnet)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(pool.Id).To(Equal(poolId))
+					Expect(pool.as.Id).To(Equal(as.Id))
+					Expect(pool.IfName).To(Equal("eth0"))
+					Expect(pool.Subnet.String()).To(Equal(poolId))
+					Expect(pool.IsIPv6).To(BeTrue())
+					Expect(pool.Priority).To(Equal(1))
+					Expect(as.Pools[poolId]).NotTo(BeNil())
+				})
+			})
+		})
+
+		Describe("Test getAddressPool", func() {
+			Context("When pool not find", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					pool, err := as.getAddressPool("pool")
+					Expect(err).To(Equal(errInvalidPoolId))
+					Expect(pool).To(BeNil())
+				})
+			})
+
+			Context("When pool is found", func() {
+				It("Should return the pool", func() {
+					poolId := "pool"
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools[poolId] = &addressPool{Id: poolId}
+					pool, err := as.getAddressPool(poolId)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(pool.Id).To(Equal(poolId))
+				})
+			})
+		})
+
+		Describe("Test requestPool", func() {
+			Context("When poolId is not empty and not found in as", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					ap, err := as.requestPool("pool", "", nil, false)
+					Expect(err).To(Equal(errAddressPoolNotFound))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When poolId is not empty and found in as", func() {
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id:       poolId,
+						RefCount: 0,
+					}
+					ap, err := as.requestPool(poolId, "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When pool is in use", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id:       poolId,
+						RefCount: 1,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When pool is ipv4 and ipv6 is wanted", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id:     poolId,
+						IsIPv6: false,
+					}
+					ap, err := as.requestPool("", "", nil, true)
```

Comment:
This test case is to test requesting an ipv6 pool from an address space which only contain ipv4 pool, by only not assign the pool id it will enter the IP version matching "if branch".